Add Notifications Support

arthurzam opened the pull request:
This Pull Request isn't finished yet in my idea, but I request your review. Please note that my schedule is very problematic, therefore If I don't reply then I can't reply at that time, but when I will have the ability I will fix and reply :smile: 

The idea in this PR is to add notifications support to QMPlay2, for example to show the title of the playing video when it moves to the next one.

Supports X11 notifications (like with notify-send) and system tray notifications.
Supports some settings changes, like format, timeout and what notifications to show. If you have more ideas of events to show notification please inform me.

Some questions for improvements:
1. Are the class names and Module name OK?
2. Is the code style fine (I checked many times to match yours but maybe I missed some place)?
3. Windows supports need testing (I can't compile now to test over Windows). The support should be using the system tray.
4. Should I move some code from one file to a new file (for example look at `Notifys.cpp` & `Notifys.hpp`)?
5. Use flags: should I add use flags (in CMake) for `USE_NOTIFY` and `USE_DBUS` (about this one I know that many users don't want dbus at all and therefore remove X11 notifications)?

zaps166 commented:
Thanks for new module :)

Typo: `Notifys`->`Notifies`

I'll check it soon :smile:

arthurzam commented:
- Fixed this typo.
- Rebased over new changes.

I think future improvements are adding more notifications services. Do you know about OS X?

zaps166 commented on src/gui/MainWidget.cpp:
```diff
@@ -154,6 +154,7 @@ MainWidget::MainWidget(QPair<QStringList, QStringList> &QMPArguments)
 	tray = new QSystemTrayIcon(this);
 	tray->setIcon(QMPlay2Core.getIconFromTheme("QMPlay2-panel", QMPlay2Core.getQMPlay2Pixmap()));
 	tray->setVisible(settings.getBool("TrayVisible", true));
+	QApplication::instance()->setProperty("system_tray", QVariant::fromValue<void *>(tray));
```
Is `QApplication::instance()` better than `qApp` macro?
I think that you can create another variable for tray in `QMPlay2Core` class (without including `QSystemTrayIcon`). API/ABI compatibility is now not important.

zaps166 commented on src/modules/Notifies/Notifies.cpp:
```diff
@@ -0,0 +1,175 @@
+Notifies::Notifies() :
+	Module("Notifies")
+{
+//	moduleImg = QImage(":/???"); // TODO: add icon
```
Please ask "varlesh" for an icon :)

zaps166 commented on src/modules/Notifies/Notifies.cpp:
```diff
@@ -0,0 +1,175 @@
+ModuleSettingsWidget::ModuleSettingsWidget(Module &module) :
+	Module::SettingsWidget(module)
+{
+	/**/
```
?

zaps166 commented on src/modules/Notifies/Notifies.pro:
```diff
@@ -0,0 +1,34 @@
+
+unix:!macx:!android {
+	QT += dbus
+        HEADERS += X11_Notify.hpp
```
tabs

zaps166 commented on src/modules/Notifies/X11_Notify.cpp:
```diff
@@ -0,0 +1,146 @@
+#if Q_BYTE_ORDER == Q_LITTLE_ENDIAN
+	// ABGR -> ARGB
+	const QImage i = scaled.rgbSwapped();
+#else
```
QMPlay2 doesn't support BigEndian CPUs. Do you have something (Qemu or real hardware) to test it?

zaps166 commented on src/modules/Notifies/X11_Notify.hpp:
```diff
@@ -0,0 +1,44 @@
+class OrgFreedesktopNotificationsInterface;
+class QDBusPendingCallWatcher;
+
+class X11_Notify : public Notify
```
Maybe `X11Notify`?
I prefer using camelCase instead of "_". I know that QMPlay2 has mixed styles, but new code has always `camelCase` for variables, `CamelCase` for classes name and `m_camelCase` for private/protected class members.
It can be applied also for INI (settings) keys (start e.g. from capital letter).

zaps166 commented on src/gui/Main.cpp:
```diff
@@ -166,7 +166,7 @@ const QWidget *QMPlay2GUIClass::getVideoDock() const
 	groupIcon(NULL), mediaIcon(NULL), folderIcon(NULL),
 	mainW(NULL),
-	screenSaver(NULL)
+	screenSaver(NULL), shortcutHandler(NULL)
```
Unrelated change, I'll do it in `master` branch.

zaps166 commented on src/modules/Notifies/NotifyExtension.cpp:
```diff
@@ -0,0 +1,119 @@
+bool NotifyExtension::set()
+{
+	if (sets().getBool("type_disabled"))
+		delete this->notify;
```
I think we should avoid `this->`. Also for new code please use `m_` prefix for private/protected class members :)

zaps166 commented on src/modules/Notifies/NotifyExtension.cpp:
```diff
@@ -0,0 +1,119 @@
+	if (settings.getBool("show_volume"))
+		connect(&QMPlay2Core, SIGNAL(volumeChanged(double)), this, SLOT(volumeChanged(double)));
+
+	if(settings.getBool("custom_msg"))
```
Always spaces after `if`, `while`, `for`, etc.

zaps166 commented:
> Do you know about OS X?

Hmm, no :smile: Don't do for OS X (macOS), probably nobody uses QMPlay2 on macOS. It also has poor integration with it.

zaps166 commented:
Also it is good to keep cpp and hpp in the same order including line breaks after function name, e.g. if header doesn't have line break between methods, cpp also shouldn't have it. Is it only my coding style ? :smile: ?

arthurzam commented:
I rebased and squashed all the commits (those renames just made my life hard when rebasing, so to fix them I rebased).

> Please ask "varlesh" for an icon :)

I think it is better that you ask him for an icon (as you are the main developer)

zaps166 commented:
> I think it is better that you ask him for an icon (as you are the main developer)

Right :)

zaps166 commented on src/modules/Notifies/TrayNotify.cpp:
```diff
@@ -0,0 +1,44 @@
+		return false;
+
+	QSystemTrayIcon *tray = QMPlay2Core.systemTray;
+
```
```c++
if (!tray)
    return false;
````

arthurzam commented:
Fixed.

> I can see Paused notification, but I can't see notification that playback is started again - is it intended?

Yes it is. There is no need to show "Playing" when there is a notification for the playing item (Do you have a better name for the playing item in playlist? ).

On another subject, I think it is a good idea to add a new Audio Writer for `OSS` (for `*BSD` systems). I can try to implement, but what are your thoughts?

zaps166 commented:
> On another subject, I think it is a good idea to add a new Audio Writer for OSS (for *BSD systems). I can try to implement, but what are your thoughts?

Who uses QMPlay2 on *BSD...? Currently they have PortAudio which it is not bad (I think) :) It makes sense if you want to use it or improve your programming skills. AFAIK OSS is very easy to implement, so you shouldn't have problems. Of course I'll review, test and merge (please open a new issue if you want to do it) :)

> There is no need to show "Playing" when there is a notification for the playing item.

Hmm, yes. But after pasued state shouldn't be shown again? Currently I can see only Paused if I'm doing sequence: pause, unpause, pause, unpause, etc...

QMPlay2 shows also 'Stopped' notification when closing and is already stopped. The same when clicking many times on stop button - maybe read/remember current state?

> Do you have a better name for the playing item in playlist?

So which name is better than "playing" ? I don't know, maybe I should ask someone who know English better than me :smile:

arthurzam commented:
> Hmm, yes. But after pasued state shouldn't be shown again? Currently I can see only Paused if I'm doing sequence: pause, unpause, pause, unpause, etc...
> 
> QMPlay2 shows also 'Stopped' notification when closing and is already stopped. The same when clicking many times on stop button - maybe read/remember current state?

Done (please check the code style in this commit).

zaps166 commented on src/modules/Notifies/NotifyExtension.hpp:
```diff
@@ -33,6 +33,11 @@ private slots:
 private:
 	Notify *m_notify;
 	QString m_summaryFormat, m_bodyFormat;
+	enum PlayState{
```
`PlayState {`

zaps166 commented on src/modules/Notifies/NotifyExtension.cpp:
```diff
@@ -84,7 +85,22 @@ void NotifyService::updatePlaying(bool play, const QString &title, const QString
+	 *  1. The last one is the same as the current one
+	 *  2. The current one is Playing and the last one wasn't Paused
+	 */
+	if (
```
This line is not so long:
```c++
if (m_lastPlayState != last && (m_lastPlayState != Playing || last == Paused))
	m_notify->showMessage(QCoreApplication::applicationName(), playState);
```

arthurzam commented:
OK, no problem :smile:
Fixed.

zaps166 commented on src/modules/Notifies/NotifyExtension.cpp:
```diff
@@ -0,0 +1,138 @@
+
+/**/
+
+NotifyExtension::NotifyExtension(Module &module) : m_notify(NULL)
```
```c++
NotifyExtension::NotifyExtension(Module &module) :
	m_notify(NULL)
{
```

Also maybe rename to `m_notifyService` (because not the same as in class above)?
And no need to line break between functions (consistent to header, if you have a time you can check all code for this) :D

zaps166 commented on src/modules/Notifies/NotifyExtension.cpp:
```diff
@@ -0,0 +1,138 @@
+
+NotifyExtension::~NotifyExtension()
+{
+	delete m_notify;
```
Open QMPlay2 when notifications are enabled. Then go to settings, disable it, apply and close QMPlay2 - crash!

zaps166 commented on src/modules/Notifies/X11Notify.cpp:
```diff
@@ -0,0 +1,147 @@
+	// ABGR -> ARGB
+	const QImage i = scaled.rgbSwapped();
+#else
+	// ABGR -> GBAR
```
Is it tested, does it make any sense (QMPlay2 doesn't support big endian - never tested)?

zaps166 commented on src/modules/Notifies/NotifyExtension.cpp:
```diff
@@ -0,0 +1,138 @@
+bool NotifyExtension::set()
+{
+	if (sets().getBool("TypeDisabled"))
+		delete m_notify;
```
Set it to `NULL` to prevent crash described above :)

zaps166 commented on src/modules/Notifies/NotifyExtension.hpp:
```diff
@@ -0,0 +1,56 @@
+{
+	Q_OBJECT
+public:
+	NotifyService(Notify *m_notify, Settings &settings);
```
`m_notify` not a class memeber here :)

zaps166 commented on src/modules/Notifies/Notify.hpp:
```diff
@@ -0,0 +1,37 @@
+public:
+	virtual bool showMessage(const QString &summary, const QString &message = QString(), const QString &icon = QString(), const QImage &image = QImage()) = 0;
+
+	virtual ~Notify() { }
```
virtual ~Notify()
{}

zaps166 commented on src/modules/Notifies/Notify.hpp:
```diff
@@ -0,0 +1,37 @@
+
+	virtual ~Notify() { }
+protected:
+	Notify(qint32 timeout) : m_timeout(timeout) {}
```
The same as something else, new line after `:` and new line before `{}` :)

zaps166 commented on src/modules/Notifies/X11Notify.cpp:
```diff
@@ -0,0 +1,147 @@
+
+#define X11NotifyName "X11 Notify"
+
+QDBusArgument &operator<<(QDBusArgument &arg, const QImage &image)
```
I can compile & run without this, hmm... Is it necessary?

arthurzam replied:
> I can compile & run without this, hmm... Is it necessary?

In previous versions of Qt it was needed to convert properly `QImage` -> dbus information. For now we don't send images at all in QMPlay2 (we send only text), but for future needs.

zaps166 commented on src/modules/Notifies/X11Notify.cpp:
```diff
@@ -0,0 +1,147 @@
+	QDBusPendingReply<uint> reply = m_interface->Notify(
+				QCoreApplication::applicationName(), id, icon, summary,
+				message, QStringList(), hints, m_timeout);
+	QDBusPendingCallWatcher* watcher = new QDBusPendingCallWatcher(reply, this);
```
Start to the right :D

arthurzam replied:
> Start to the right

Sorry, but I didn't understand.

zaps166 replied:
`QDBusPendingCallWatcher* watcher` -> `QDBusPendingCallWatcher *watcher`, but not important

zaps166 commented on src/modules/Notifies/X11Notify.cpp:
```diff
@@ -0,0 +1,147 @@
+	Notify(timeout),
+	m_notificationId(0)
+{
+	m_interface = new OrgFreedesktopNotificationsInterface(
```
I think that no need to split line in QMPlay2. Splitting line using only tabs (which QMPlay2 uses) is not good :D

zaps166 commented:
This contains an `X11Notify` class, but it uses freedesktop notifications via D-Bus. If it can be used outside X11, so maybe rename it to `FreedesktopNotify` ?

arthurzam commented:
Fixed :smile:

zaps166 commented:
:+1: Now it works much better when it reuses a notification :)

zaps166 commented:
Thanks!
Anything else can be fixed by separate commit/PR :D
